### PR TITLE
[doc] Add some new ADRs on the evolution of our layout algorithm

### LINKED ISSUE #1755: Contribute a specification of the behavior of the layout

### PR DESCRIPTION
**Pull request by sbegaudeau**
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/094_separate_the_layout_data_structures_from_the_diagram_data_structures_.adoc [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+We would simply have to copy the diagram with its new layout data:
+
+```
+var layoutedDiagram = Diagram.newDiagram(nonLayoutedDiagram)
+    .layoutData(layoutData)
+    .build();
+```
+
+These new data structures will simply allow us to manipulate more easily all the layout data of a diagram.
+If new nodes or new edges have been created during a new rendering, then they should not have any associated layout data.
```
Comment (AxelRICHARD) [inline doc/adrs/094_separate_the_layout_data_structures_from_the_diagram_data_structures_.adoc]: Why new nodes an edges should not have layout data? If they are visible in the diagram, they should have layout data?

File: doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+For example, if a child node is located in a large parent node with a list layout, this will not change the size of the border of the child node.
+
+
+==== Footprint and constraints
+
+In some use case, to match a footprint requested by our parent, we will have to compute another footprint for a node.
+The configuration of a node should indicate which boxes should be resized in order to match the newly requested footprint.
+
+For example, if we consider a node with a list layout in a column, it may ask some of its children to occupy a larger footprint than necessary.
+In one case, the specifier may want those children to expand their outside label box to occupy this larger footprint, for example for the attributes of a class diagram.
```
Comment (AxelRICHARD) [inline doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc]: I don't understand why you are referring outside label box here, but in the associated picture below, it seems inside label boxes are used.

File: doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+If the newly moved nodes are also overlapping other nodes, we will move these nodes in one direction to have enough space.
+We will continue to do so recursively.
+Since a diagram has no bounds in any direction and since a node can be as big as necessary, moving existing nodes in this way can be done recursively without issues.
+
+NOTE: Insert a demo
+
+=== Inside label box
+
+The inside label box is a place where labels can be put inside of the node.
+These labels will count torward the size and thus the footprint of the node.
```
Comment (AxelRICHARD) [inline doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc]: torward -> toward

File: doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+Child nodes can only be layouted with a position inside the content box.
+After child nodes have been layouted, the full size of the content box will be known.
+Then we will compute the size of the label box, as a result a child node will never be able to be placed in the inside label box.
```
Comment (AxelRICHARD) [inline doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc]: Shouldn't it be the inverse? (label box computed before the child nodes)

File: doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+
+
+== Status
+
+Work in progress
+
+
+== Consequences
+
```
Comment (AxelRICHARD) [inline doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc]: The ELK configuration will not be used anymore by this new box model, right? If it is the case, could you add it to this ADR please?

File: doc/adrs/094_separate_the_layout_data_structures_from_the_diagram_data_structures_.adoc (hunk 1 above)
Comment (pcdavid) [reply doc/adrs/094_separate_the_layout_data_structures_from_the_diagram_data_structures_.adoc]: I think what Stéphane meant is that the diagram component rendering (*before* the layout is called, as the actual layout is not in the scope of this ADR), will simply not create "dummy" layout data for newly created elements. Currently the layout data is stored in e.g. `Node.size` which is non-nullable, so `NodeComponent.render()` *must* provide some value for new nodes, and we end up with `Size#UNDEFINED` as mentioned below which needs explicit handling everywhere.

Producing actual meaningfull layout information for newly created elements will be the job of the layout engine, not the diagram renderer which will focus on the logical structure of the diagram and determining the effective style applicable to each element.

File: doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+When the project was first started a couple years ago, we only used ELK and a fully automatic layout.
+After that, we wanted to let end users move nodes by themselves and so we introduced the incremental layout.
+
+This incremental layout works by having an existing layout and the current event in order to compute a new layout.
+It was a very simple approach which produced good results for a low cost.
+
+It does not really know how to layout everything from scratch, which is why everything is layouted with ELK first, it simply determines how an event impact the current layout.
+One of the issue with this approach is that since we have more and more complex diagrams, we need to handle a very complex set of interactions between an existing layout and all the possible events.
+On top of that, we also had to deal with complex unsynchronized diagrams which can start with a blank state.
+In those case, ELK will just layout an empty diagram and the incremental layout will have to handle everything manually.
```
Comment (pcdavid) [inline doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc]: "manually" => "by itself".

It's not *that* ambiguous in the context, but "manually" could be interpreted as "the end-user has to layout every element by himself".

File: doc/adrs/094_separate_the_layout_data_structures_from_the_diagram_data_structures_.adoc (hunk 1 above)
Comment (sbegaudeau) [reply doc/adrs/094_separate_the_layout_data_structures_from_the_diagram_data_structures_.adoc]: Exactly, DiagramRender#render will return a Diagram data structure without any meaningful layout data.

File: doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc (hunk 4 above)
Comment (sbegaudeau) [reply doc/adrs/095_simplify_the_manipulation_of_nodes_by_switching_to_a_common_box_model.adoc]: The label box is encapsulating the content box (which contains the child nodes). So we will compute the content box and after that, when we will have its size, we will compute the label box which will surround it

**Review by pcdavid — APPROVED** [review]

**Pull request merged**